--- before/net/minecraft/entity/projectile/EntityLargeFireball.java
+++ after/net/minecraft/entity/projectile/EntityLargeFireball.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.projectile;
 
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.DamageSource;
@@ -8,6 +9,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class EntityLargeFireball extends EntityFireball
 {
@@ -16,12 +18,12 @@
     public EntityLargeFireball(World p_i1767_1_)
     {
         super(p_i1767_1_);
+        // TODO: Maybe we should use net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent instead of this?
+        isIncendiary = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
     }
 
     @SideOnly(Side.CLIENT)
-    public EntityLargeFireball(
-        World p_i1768_1_, double p_i1768_2_, double p_i1768_4_, double p_i1768_6_, double p_i1768_8_, double p_i1768_10_, double p_i1768_12_
-    )
+    public EntityLargeFireball(World p_i1768_1_, double p_i1768_2_, double p_i1768_4_, double p_i1768_6_, double p_i1768_8_, double p_i1768_10_, double p_i1768_12_)
     {
         super(p_i1768_1_, p_i1768_2_, p_i1768_4_, p_i1768_6_, p_i1768_8_, p_i1768_10_, p_i1768_12_);
     }
@@ -29,9 +31,10 @@
     public EntityLargeFireball(World p_i1769_1_, EntityLivingBase p_i1769_2_, double p_i1769_3_, double p_i1769_5_, double p_i1769_7_)
     {
         super(p_i1769_1_, p_i1769_2_, p_i1769_3_, p_i1769_5_, p_i1769_7_);
+        // TODO: Maybe we should use net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent instead of this?
+        isIncendiary = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
     }
 
-    @Override
     protected void func_70227_a(RayTraceResult p_70227_1_)
     {
         if (!this.field_70170_p.field_72995_K)
@@ -42,8 +45,17 @@
                 this.func_174815_a(this.field_70235_a, p_70227_1_.field_72308_g);
             }
 
-            boolean flag = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
-            this.field_70170_p.func_72885_a(null, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_92057_e, flag, flag);
+            // TODO: Reimplement with correct `flag` usage below
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a);
+            // this.world.newExplosion((Entity)null, this.posX, this.posY, this.posZ, (float)this.explosionPower, flag, flag);
+            // CraftBukkit start - fire ExplosionPrimeEvent
+            ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) org.bukkit.craftbukkit.v1_12_R1.entity.CraftEntity.getEntity(this.field_70170_p.getServer(), this));
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                // give 'this' instead of (Entity) null so we know what causes the damage
+                this.field_70170_p.func_72885_a(this, this.field_70165_t, this.field_70163_u, this.field_70161_v, event.getRadius(), event.getFire(), isIncendiary);
+            }
+            // CraftBukkit end
             this.func_70106_y();
         }
     }
@@ -53,21 +65,20 @@
         EntityFireball.func_189743_a(p_189744_0_, "Fireball");
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("ExplosionPower", this.field_92057_e);
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
 
         if (p_70037_1_.func_150297_b("ExplosionPower", 99))
         {
-            this.field_92057_e = p_70037_1_.func_74762_e("ExplosionPower");
+            // CraftBukkit - set bukkitYield when setting explosionpower
+            bukkitYield = this.field_92057_e = p_70037_1_.func_74762_e("ExplosionPower");
         }
     }
 }
