--- before/net/minecraft/world/chunk/storage/IChunkLoader.java
+++ after/net/minecraft/world/chunk/storage/IChunkLoader.java
@@ -20,4 +20,10 @@
     void func_75818_b();
 
     boolean func_191063_a(int p_191063_1_, int p_191063_2_);
+
+    // CatServer start - Add spigot method
+    default void saveChunk(World world, Chunk chunk, boolean unloaded) throws MinecraftException, IOException {
+        func_75816_a(world, chunk);
+    }
+    // CatServer end
 }
