--- before/net/minecraft/item/ItemLilyPad.java
+++ after/net/minecraft/item/ItemLilyPad.java
@@ -25,7 +25,6 @@
         super(p_i45357_1_, false);
     }
 
-    @Override
     public ActionResult<ItemStack> func_77659_a(World p_77659_1_, EntityPlayer p_77659_2_, EnumHand p_77659_3_)
     {
         ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
@@ -33,7 +32,7 @@
 
         if (raytraceresult == null)
         {
-            return new ActionResult<>(EnumActionResult.PASS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
         }
         else
         {
@@ -41,21 +40,35 @@
             {
                 BlockPos blockpos = raytraceresult.func_178782_a();
 
-                if (!p_77659_1_.func_175660_a(p_77659_2_, blockpos)
-                        || !p_77659_2_.func_175151_a(blockpos.func_177972_a(raytraceresult.field_178784_b), raytraceresult.field_178784_b, itemstack))
+                if (!p_77659_1_.func_175660_a(p_77659_2_, blockpos) || !p_77659_2_.func_175151_a(blockpos.func_177972_a(raytraceresult.field_178784_b), raytraceresult.field_178784_b, itemstack))
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
 
                 BlockPos blockpos1 = blockpos.func_177984_a();
                 IBlockState iblockstate = p_77659_1_.func_180495_p(blockpos);
 
-                if (iblockstate.func_185904_a() == Material.field_151586_h
-                        && iblockstate.func_177229_b(BlockLiquid.field_176367_b) == 0
-                        && p_77659_1_.func_175623_d(blockpos1))
+                if (iblockstate.func_185904_a() == Material.field_151586_h && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0 && p_77659_1_.func_175623_d(blockpos1))
                 {
+                    // CraftBukkit start - special case for handling block placement with water lilies
+                    org.bukkit.block.BlockState blockstate = org.bukkit.craftbukkit.v1_12_R1.block.CraftBlockState.getBlockState(p_77659_1_, blockpos1.func_177958_n(), blockpos1.func_177956_o(), blockpos1.func_177952_p());
+                    // special case for handling block placement with water lilies
+                    net.minecraftforge.common.util.BlockSnapshot blocksnapshot = net.minecraftforge.common.util.BlockSnapshot.getBlockSnapshot(p_77659_1_, blockpos1);
+                    p_77659_1_.func_175656_a(blockpos1, Blocks.field_150392_bi.func_176223_P());
+                    if (net.minecraftforge.event.ForgeEventFactory.onPlayerBlockPlace(p_77659_2_, blocksnapshot, net.minecraft.util.EnumFacing.UP, p_77659_3_).isCanceled())
+                    {
+                        blocksnapshot.restore(true, false);
+                        return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
+                    }
+
                     p_77659_1_.func_180501_a(blockpos1, Blocks.field_150392_bi.func_176223_P(), 11);
 
+                    org.bukkit.event.block.BlockPlaceEvent placeEvent = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callBlockPlaceEvent(p_77659_1_, p_77659_2_, p_77659_3_, blockstate, blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p());
+                    if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild())) {
+                        blockstate.update(true, false);
+                        return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                    }
+
                     if (p_77659_2_ instanceof EntityPlayerMP)
                     {
                         CriteriaTriggers.field_193137_x.func_193173_a((EntityPlayerMP)p_77659_2_, blockpos1, itemstack);
@@ -68,11 +81,11 @@
 
                     p_77659_2_.func_71029_a(StatList.func_188057_b(this));
                     p_77659_1_.func_184133_a(p_77659_2_, blockpos, SoundEvents.field_187916_gp, SoundCategory.BLOCKS, 1.0F, 1.0F);
-                    return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
                 }
             }
 
-            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
         }
     }
 }
