--- before/net/minecraft/block/BlockDispenser.java
+++ after/net/minecraft/block/BlockDispenser.java
@@ -132,7 +132,7 @@
         }
     }
 
-    protected void func_176439_d(World p_176439_1_, BlockPos p_176439_2_)
+    public void func_176439_d(World p_176439_1_, BlockPos p_176439_2_)
     {
         BlockSourceImpl blocksourceimpl = new BlockSourceImpl(p_176439_1_, p_176439_2_);
         TileEntityDispenser tileentitydispenser = blocksourceimpl.func_150835_j();
