--- before/net/minecraft/block/BlockSilverfish.java
+++ after/net/minecraft/block/BlockSilverfish.java
@@ -13,6 +13,7 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class BlockSilverfish extends Block
 {
@@ -66,10 +67,8 @@
         if (!p_180653_1_.field_72995_K && p_180653_1_.func_82736_K().func_82766_b("doTileDrops"))
         {
             EntitySilverfish entitysilverfish = new EntitySilverfish(p_180653_1_);
-            entitysilverfish.func_70012_b(
-                (double)p_180653_2_.func_177958_n() + 0.5, (double)p_180653_2_.func_177956_o(), (double)p_180653_2_.func_177952_p() + 0.5, 0.0F, 0.0F
-            );
-            p_180653_1_.func_72838_d(entitysilverfish);
+            entitysilverfish.func_70012_b((double)p_180653_2_.func_177958_n() + 0.5D, (double)p_180653_2_.func_177956_o(), (double)p_180653_2_.func_177952_p() + 0.5D, 0.0F, 0.0F);
+            p_180653_1_.addEntity(entitysilverfish, CreatureSpawnEvent.SpawnReason.SILVERFISH_BLOCK); // CraftBukkit - add SpawnReason
             entitysilverfish.func_70656_aK();
         }
     }
