--- before/net/minecraft/entity/monster/EntitySkeleton.java
+++ after/net/minecraft/entity/monster/EntitySkeleton.java
@@ -62,7 +62,7 @@
     @Override
     public void func_70645_a(DamageSource p_70645_1_)
     {
-        super.func_70645_a(p_70645_1_);
+        // super.onDeath(cause); // CraftBukkit - moved down
 
         if (p_70645_1_.func_76346_g() instanceof EntityCreeper)
         {
@@ -74,6 +74,7 @@
                 this.func_70099_a(new ItemStack(Items.field_151144_bL, 1, 0), 0.0F);
             }
         }
+        super.func_70645_a(p_70645_1_); // CraftBukkit - moved from above
     }
 
     @Override
