--- before/net/minecraft/entity/EntityTrackerEntry.java
+++ after/net/minecraft/entity/EntityTrackerEntry.java
@@ -66,6 +66,10 @@
 import net.minecraft.world.storage.MapData;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerVelocityEvent;
+
+import javax.annotation.Nullable;
 
 public class EntityTrackerEntry
 {
@@ -90,11 +94,11 @@
     private boolean field_73144_s;
     private final boolean field_73143_t;
     private int field_73142_u;
-    private List<Entity> field_187263_w = Collections.emptyList();
+    private List<Entity> field_187263_w = Collections.<Entity>emptyList();
     private boolean field_73141_v;
     private boolean field_180234_y;
     public boolean field_73133_n;
-    public final Set<EntityPlayerMP> field_73134_o = Sets.newHashSet();
+    public final Set<EntityPlayerMP> field_73134_o = Sets.<EntityPlayerMP>newHashSet();
 
     public EntityTrackerEntry(Entity p_i46837_1_, int p_i46837_2_, int p_i46837_3_, int p_i46837_4_, boolean p_i46837_5_)
     {
@@ -112,7 +116,6 @@
         this.field_180234_y = p_i46837_1_.field_70122_E;
     }
 
-    @Override
     public boolean equals(Object p_equals_1_)
     {
         if (p_equals_1_ instanceof EntityTrackerEntry)
@@ -125,7 +128,6 @@
         }
     }
 
-    @Override
     public int hashCode()
     {
         return this.field_73132_a.func_145782_y();
@@ -135,7 +137,7 @@
     {
         this.field_73133_n = false;
 
-        if (!this.field_73144_s || this.field_73132_a.func_70092_e(this.field_73147_p, this.field_73146_q, this.field_73145_r) > 16.0)
+        if (!this.field_73144_s || this.field_73132_a.func_70092_e(this.field_73147_p, this.field_73146_q, this.field_73145_r) > 16.0D)
         {
             this.field_73147_p = this.field_73132_a.field_70165_t;
             this.field_73146_q = this.field_73132_a.field_70163_u;
@@ -153,20 +155,20 @@
             this.func_151259_a(new SPacketSetPassengers(this.field_73132_a));
         }
 
-        if (this.field_73132_a instanceof EntityItemFrame && this.field_73136_m % 10 == 0)
+        if (this.field_73132_a instanceof EntityItemFrame /*&& this.updateCounter % 10 == 0*/) // CraftBukkit - Moved below, should always enter this block
         {
             EntityItemFrame entityitemframe = (EntityItemFrame)this.field_73132_a;
             ItemStack itemstack = entityitemframe.func_82335_i();
 
-            if (itemstack.func_77973_b() instanceof ItemMap)
+            if (this.field_73136_m % 10 == 0 && itemstack.func_77973_b() instanceof ItemMap)// CraftBukkit - Moved this.updateCounter % 10 logic here so item frames do not enter the other blocks
             {
-                MapData mapdata = Items.field_151098_aY.func_77873_a(itemstack, this.field_73132_a.field_70170_p);
+                MapData mapdata = ((ItemMap) itemstack.func_77973_b()).func_77873_a(itemstack, this.field_73132_a.field_70170_p);
 
-                for (EntityPlayer entityplayer : p_73122_1_)
+                for (EntityPlayer entityplayer : field_73134_o)
                 {
                     EntityPlayerMP entityplayermp = (EntityPlayerMP)entityplayer;
                     mapdata.func_76191_a(entityplayermp, itemstack);
-                    Packet<?> packet = Items.field_151098_aY.func_150911_c(itemstack, this.field_73132_a.field_70170_p, entityplayermp);
+                    Packet<?> packet = ((ItemMap) itemstack.func_77973_b()).func_150911_c(itemstack, this.field_73132_a.field_70170_p, entityplayermp);
 
                     if (packet != null)
                     {
@@ -188,9 +190,7 @@
 
                 if (flag3)
                 {
-                    this.func_151259_a(
-                        new SPacketEntity.S16PacketEntityLook(this.field_73132_a.func_145782_y(), (byte)j1, (byte)l1, this.field_73132_a.field_70122_E)
-                    );
+                    this.func_151259_a(new SPacketEntity.S16PacketEntityLook(this.field_73132_a.func_145782_y(), (byte)j1, (byte)l1, this.field_73132_a.field_70122_E));
                     this.field_73127_g = j1;
                     this.field_73139_h = l1;
                 }
@@ -215,18 +215,23 @@
                 Packet<?> packet1 = null;
                 boolean flag = j * j + k * k + l * l >= 128L || this.field_73136_m % 60 == 0;
                 boolean flag1 = Math.abs(k2 - this.field_73127_g) >= 1 || Math.abs(i - this.field_73139_h) >= 1;
+                // CraftBukkit start - Code moved from below
+                if (flag)
+                {
+                    this.field_73128_d = i1;
+                    this.field_73129_e = i2;
+                    this.field_73126_f = j2;
+                }
 
+                if (flag1)
+                {
+                    this.field_73127_g = k2;
+                    this.field_73139_h = i;
+                }
+                // CraftBukkit end
                 if (this.field_73136_m > 0 || this.field_73132_a instanceof EntityArrow)
                 {
-                    if (j >= -32768L
-                            && j < 32768L
-                            && k >= -32768L
-                            && k < 32768L
-                            && l >= -32768L
-                            && l < 32768L
-                            && this.field_73142_u <= 400
-                            && !this.field_73141_v
-                            && this.field_180234_y == this.field_73132_a.field_70122_E)
+                    if (j >= -32768L && j < 32768L && k >= -32768L && k < 32768L && l >= -32768L && l < 32768L && this.field_73142_u <= 400 && !this.field_73141_v && this.field_180234_y == this.field_73132_a.field_70122_E)
                     {
                         if ((!flag || !flag1) && !(this.field_73132_a instanceof EntityArrow))
                         {
@@ -241,15 +246,18 @@
                         }
                         else
                         {
-                            packet1 = new SPacketEntity.S17PacketEntityLookMove(
-                                this.field_73132_a.func_145782_y(), j, k, l, (byte)k2, (byte)i, this.field_73132_a.field_70122_E
-                            );
+                            packet1 = new SPacketEntity.S17PacketEntityLookMove(this.field_73132_a.func_145782_y(), j, k, l, (byte)k2, (byte)i, this.field_73132_a.field_70122_E);
                         }
                     }
                     else
                     {
                         this.field_180234_y = this.field_73132_a.field_70122_E;
                         this.field_73142_u = 0;
+                        // CraftBukkit start - Refresh list of who can see a player before sending teleport packet
+                        if (this.field_73132_a instanceof EntityPlayer) {
+                            this.func_73125_b(new java.util.ArrayList<>(this.field_73134_o));
+                        }
+                        // CraftBukkit end
                         this.func_187261_c();
                         packet1 = new SPacketEntityTeleport(this.field_73132_a);
                     }
@@ -267,11 +275,10 @@
                     double d0 = this.field_73132_a.field_70159_w - this.field_73137_j;
                     double d1 = this.field_73132_a.field_70181_x - this.field_73138_k;
                     double d2 = this.field_73132_a.field_70179_y - this.field_73135_l;
-                    double d3 = 0.02;
+                    double d3 = 0.02D;
                     double d4 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                    if (d4 > 4.0E-4
-                            || d4 > 0.0 && this.field_73132_a.field_70159_w == 0.0 && this.field_73132_a.field_70181_x == 0.0 && this.field_73132_a.field_70179_y == 0.0)
+                    if (d4 > 4.0E-4D || d4 > 0.0D && this.field_73132_a.field_70159_w == 0.0D && this.field_73132_a.field_70181_x == 0.0D && this.field_73132_a.field_70179_y == 0.0D)
                     {
                         this.field_73137_j = this.field_73132_a.field_70159_w;
                         this.field_73138_k = this.field_73132_a.field_70181_x;
@@ -287,19 +294,20 @@
 
                 this.func_111190_b();
 
+                /* CraftBukkit start - Code moved up
                 if (flag)
                 {
-                    this.field_73128_d = i1;
-                    this.field_73129_e = i2;
-                    this.field_73126_f = j2;
+                    this.encodedPosX = i1;
+                    this.encodedPosY = i2;
+                    this.encodedPosZ = j2;
                 }
 
                 if (flag1)
                 {
-                    this.field_73127_g = k2;
-                    this.field_73139_h = i;
+                    this.encodedRotationYaw = k2;
+                    this.encodedRotationPitch = i;
                 }
-
+                // CraftBukkit end */
                 this.field_73141_v = false;
             }
 
@@ -318,7 +326,28 @@
 
         if (this.field_73132_a.field_70133_I)
         {
-            this.func_151261_b(new SPacketEntityVelocity(this.field_73132_a));
+            // this.sendToTrackingAndSelf(new SPacketEntityVelocity(this.trackedEntity));
+            // CraftBukkit start - Create PlayerVelocity event
+            boolean cancelled = false;
+
+            if (this.field_73132_a instanceof EntityPlayer) {
+                Player player = (Player) this.field_73132_a.getBukkitEntity();
+                org.bukkit.util.Vector velocity = player.getVelocity();
+
+                PlayerVelocityEvent event = new PlayerVelocityEvent(player, velocity.clone());
+                this.field_73132_a.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    cancelled = true;
+                } else if (!velocity.equals(event.getVelocity())) {
+                    player.setVelocity(event.getVelocity());
+                }
+            }
+
+            if (!cancelled) {
+                this.func_151261_b(new SPacketEntityVelocity(this.field_73132_a));
+            }
+            // CraftBukkit end
             this.field_73132_a.field_70133_I = false;
         }
     }
@@ -339,6 +368,11 @@
 
             if (!set.isEmpty())
             {
+                // CraftBukkit start - Send scaled max health
+                if (this.field_73132_a instanceof EntityPlayerMP) {
+                    ((EntityPlayerMP) this.field_73132_a).getBukkitEntity().injectScaledMaxHealth(set, false);
+                }
+                // CraftBukkit end
                 this.func_151261_b(new SPacketEntityProperties(this.field_73132_a.func_145782_y(), set));
             }
 
@@ -375,6 +409,7 @@
 
     public void func_73118_a(EntityPlayerMP p_73118_1_)
     {
+        org.spigotmc.AsyncCatcher.catchOp("player tracker clear"); // Spigot
         if (this.field_73134_o.contains(p_73118_1_))
         {
             this.field_73132_a.func_184203_c(p_73118_1_);
@@ -385,20 +420,30 @@
 
     public void func_73117_b(EntityPlayerMP p_73117_1_)
     {
+        org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
         if (p_73117_1_ != this.field_73132_a)
         {
             if (this.func_180233_c(p_73117_1_))
             {
                 if (!this.field_73134_o.contains(p_73117_1_) && (this.func_73121_d(p_73117_1_) || this.field_73132_a.field_98038_p))
                 {
+                    // CraftBukkit start - respect vanish API
+                    if (this.field_73132_a instanceof EntityPlayerMP) {
+                        Player player = ((EntityPlayerMP) this.field_73132_a).getBukkitEntity();
+                        if (!p_73117_1_.getBukkitEntity().canSee(player)) {
+                            return;
+                        }
+                    }
+
+                    p_73117_1_.field_71130_g.remove(Integer.valueOf(this.field_73132_a.func_145782_y()));
+                    // CraftBukkit end
                     this.field_73134_o.add(p_73117_1_);
                     Packet<?> packet = this.func_151260_c();
                     p_73117_1_.field_71135_a.func_147359_a(packet);
 
                     if (!this.field_73132_a.func_184212_Q().func_187228_d())
                     {
-                        p_73117_1_.field_71135_a
-                        .func_147359_a(new SPacketEntityMetadata(this.field_73132_a.func_145782_y(), this.field_73132_a.func_184212_Q(), true));
+                        p_73117_1_.field_71135_a.func_147359_a(new SPacketEntityMetadata(this.field_73132_a.func_145782_y(), this.field_73132_a.func_184212_Q(), true));
                     }
 
                     boolean flag = this.field_73143_t;
@@ -408,6 +453,14 @@
                         AttributeMap attributemap = (AttributeMap)((EntityLivingBase)this.field_73132_a).func_110140_aT();
                         Collection<IAttributeInstance> collection = attributemap.func_111160_c();
 
+                        // CraftBukkit start - If sending own attributes send scaled health instead of current maximum health
+                        if (this.field_73132_a.func_145782_y() == p_73117_1_.func_145782_y()) {
+                            // TODO: Maybe we should check for instanceof before casting to EntityPlayerMP?
+                            ((EntityPlayerMP) this.field_73132_a).getBukkitEntity().injectScaledMaxHealth(collection, false);
+                        }
+                        // CraftBukkit end
+
+
                         if (!collection.isEmpty())
                         {
                             p_73117_1_.field_71135_a.func_147359_a(new SPacketEntityProperties(this.field_73132_a.func_145782_y(), collection));
@@ -425,15 +478,7 @@
 
                     if (flag && !(packet instanceof SPacketSpawnMob))
                     {
-                        p_73117_1_.field_71135_a
-                        .func_147359_a(
-                            new SPacketEntityVelocity(
-                                this.field_73132_a.func_145782_y(),
-                                this.field_73132_a.field_70159_w,
-                                this.field_73132_a.field_70181_x,
-                                this.field_73132_a.field_70179_y
-                            )
-                        );
+                        p_73117_1_.field_71135_a.func_147359_a(new SPacketEntityVelocity(this.field_73132_a.func_145782_y(), this.field_73132_a.field_70159_w, this.field_73132_a.field_70181_x, this.field_73132_a.field_70179_y));
                     }
 
                     if (this.field_73132_a instanceof EntityLivingBase)
@@ -459,6 +504,11 @@
                         }
                     }
 
+                    // CraftBukkit start - Fix for nonsensical head yaw
+                    this.field_73140_i = MathHelper.func_76141_d(this.field_73132_a.func_70079_am() * 256.0F / 360.0F);
+                    this.func_151259_a(new SPacketEntityHeadLook(this.field_73132_a, (byte) field_73140_i));
+                    // CraftBukkit end
+
                     if (this.field_73132_a instanceof EntityLivingBase)
                     {
                         EntityLivingBase entitylivingbase = (EntityLivingBase)this.field_73132_a;
@@ -481,6 +531,7 @@
 
                     this.field_73132_a.func_184178_b(p_73117_1_);
                     p_73117_1_.func_184848_d(this.field_73132_a);
+                    net.minecraftforge.event.ForgeEventFactory.onStartEntityTracking(field_73132_a, p_73117_1_);
                 }
             }
             else if (this.field_73134_o.contains(p_73117_1_))
@@ -488,14 +539,15 @@
                 this.field_73134_o.remove(p_73117_1_);
                 this.field_73132_a.func_184203_c(p_73117_1_);
                 p_73117_1_.func_152339_d(this.field_73132_a);
+                net.minecraftforge.event.ForgeEventFactory.onStopEntityTracking(field_73132_a, p_73117_1_);
             }
         }
     }
 
     public boolean func_180233_c(EntityPlayerMP p_180233_1_)
     {
-        double d0 = p_180233_1_.field_70165_t - (double)this.field_73128_d / 4096.0;
-        double d1 = p_180233_1_.field_70161_v - (double)this.field_73126_f / 4096.0;
+        double d0 = p_180233_1_.field_70165_t - (double)this.field_73128_d / 4096.0D;
+        double d1 = p_180233_1_.field_70161_v - (double)this.field_73126_f / 4096.0D;
         int i = Math.min(this.field_73130_b, this.field_187262_f);
         return d0 >= (double)(-i) && d0 <= (double)i && d1 >= (double)(-i) && d1 <= (double)i && this.field_73132_a.func_174827_a(p_180233_1_);
     }
@@ -513,12 +565,18 @@
         }
     }
 
+    @Nullable
     private Packet<?> func_151260_c()
     {
         if (this.field_73132_a.field_70128_L)
         {
-            field_151262_p.warn("Fetching addPacket for removed entity");
+            // CraftBukkit start - Remove useless error spam, just return
+            // LOGGER.warn("Fetching addPacket for removed entity");
+            return null;
         }
+
+        Packet pkt = net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.getEntitySpawningPacket(this.field_73132_a);
+        if (pkt != null) return pkt;
 
         if (this.field_73132_a instanceof EntityPlayerMP)
         {
@@ -621,17 +679,17 @@
                 spacketspawnobject = new SPacketSpawnObject(this.field_73132_a, i, 0);
             }
 
-            spacketspawnobject.func_149003_d((int)(entityfireball.field_70232_b * 8000.0));
-            spacketspawnobject.func_149000_e((int)(entityfireball.field_70233_c * 8000.0));
-            spacketspawnobject.func_149007_f((int)(entityfireball.field_70230_d * 8000.0));
+            spacketspawnobject.func_149003_d((int)(entityfireball.field_70232_b * 8000.0D));
+            spacketspawnobject.func_149000_e((int)(entityfireball.field_70233_c * 8000.0D));
+            spacketspawnobject.func_149007_f((int)(entityfireball.field_70230_d * 8000.0D));
             return spacketspawnobject;
         }
         else if (this.field_73132_a instanceof EntityShulkerBullet)
         {
             SPacketSpawnObject spacketspawnobject1 = new SPacketSpawnObject(this.field_73132_a, 67, 0);
-            spacketspawnobject1.func_149003_d((int)(this.field_73132_a.field_70159_w * 8000.0));
-            spacketspawnobject1.func_149000_e((int)(this.field_73132_a.field_70181_x * 8000.0));
-            spacketspawnobject1.func_149007_f((int)(this.field_73132_a.field_70179_y * 8000.0));
+            spacketspawnobject1.func_149003_d((int)(this.field_73132_a.field_70159_w * 8000.0D));
+            spacketspawnobject1.func_149000_e((int)(this.field_73132_a.field_70181_x * 8000.0D));
+            spacketspawnobject1.func_149007_f((int)(this.field_73132_a.field_70179_y * 8000.0D));
             return spacketspawnobject1;
         }
         else if (this.field_73132_a instanceof EntityEgg)
