--- before/net/minecraft/client/renderer/entity/layers/LayerCustomHead.java
+++ after/net/minecraft/client/renderer/entity/layers/LayerCustomHead.java
@@ -97,7 +97,7 @@
 
                         if (!StringUtils.isBlank(s))
                         {
-                            gameprofile = TileEntitySkull.func_174884_b(new GameProfile(null, s));
+                            gameprofile = TileEntitySkull.updateGameprofile(new GameProfile(null, s));
                             nbttagcompound.func_74782_a("SkullOwner", NBTUtil.func_180708_a(new NBTTagCompound(), gameprofile));
                         }
                     }
