--- before/net/minecraft/entity/passive/EntityLlama.java
+++ after/net/minecraft/entity/passive/EntityLlama.java
@@ -65,7 +65,7 @@
         this.func_70105_a(0.9F, 1.87F);
     }
 
-    private void func_190706_p(int p_190706_1_)
+    public void func_190706_p(int p_190706_1_)
     {
         this.field_70180_af.func_187227_b(field_190720_bG, Math.max(1, Math.min(5, p_190706_1_)));
     }
@@ -488,6 +488,9 @@
     @Override
     public void func_180430_e(float p_180430_1_, float p_180430_2_)
     {
+        float[] ret = net.minecraftforge.common.ForgeHooks.onLivingFall(this, p_180430_1_, p_180430_2_);
+        if (ret == null) return;
+        p_180430_1_ = ret[0]; p_180430_2_ = ret[1];
         int i = MathHelper.func_76123_f((p_180430_1_ * 0.5F - 3.0F) * p_180430_2_);
 
         if (i > 0)
