--- before/net/minecraft/entity/ai/EntityAISkeletonRiders.java
+++ after/net/minecraft/entity/ai/EntityAISkeletonRiders.java
@@ -1,6 +1,7 @@
 package net.minecraft.entity.ai;
 
 import net.minecraft.enchantment.EnchantmentHelper;
+import net.minecraft.entity.IEntityLivingData;
 import net.minecraft.entity.effect.EntityLightningBolt;
 import net.minecraft.entity.monster.EntitySkeleton;
 import net.minecraft.entity.passive.AbstractHorse;
@@ -11,6 +12,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.DifficultyInstance;
 
+import javax.annotation.Nullable;
+
 public class EntityAISkeletonRiders extends EntityAIBase
 {
     private final EntitySkeletonHorse field_188516_a;
@@ -43,17 +46,19 @@
             )
         );
         EntitySkeleton entityskeleton = this.func_188514_a(difficultyinstance, this.field_188516_a);
-        entityskeleton.func_184220_m(this.field_188516_a);
+        if (entityskeleton != null) entityskeleton.func_184220_m(this.field_188516_a);
 
         for (int i = 0; i < 3; ++i)
         {
             AbstractHorse abstracthorse = this.func_188515_a(difficultyinstance);
+            if (abstracthorse == null) continue; // CraftBukkit
             EntitySkeleton entityskeleton1 = this.func_188514_a(difficultyinstance, abstracthorse);
-            entityskeleton1.func_184220_m(abstracthorse);
-            abstracthorse.func_70024_g(this.field_188516_a.func_70681_au().nextGaussian() * 0.5, 0.0, this.field_188516_a.func_70681_au().nextGaussian() * 0.5);
+            if (entityskeleton1 != null) entityskeleton1.func_184220_m(abstracthorse);
+            abstracthorse.func_70024_g(this.field_188516_a.func_70681_au().nextGaussian() * 0.5D, 0.0D, this.field_188516_a.func_70681_au().nextGaussian() * 0.5D);
         }
     }
 
+    @Nullable
     private AbstractHorse func_188515_a(DifficultyInstance p_188515_1_)
     {
         EntitySkeletonHorse entityskeletonhorse = new EntitySkeletonHorse(this.field_188516_a.field_70170_p);
@@ -63,10 +68,11 @@
         entityskeletonhorse.func_110163_bv();
         entityskeletonhorse.func_110234_j(true);
         entityskeletonhorse.func_70873_a(0);
-        entityskeletonhorse.field_70170_p.func_72838_d(entityskeletonhorse);
+        if (!entityskeletonhorse.field_70170_p.addEntity(entityskeletonhorse, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.TRAP)) return null; // CraftBukkit
         return entityskeletonhorse;
     }
 
+    @Nullable
     private EntitySkeleton func_188514_a(DifficultyInstance p_188514_1_, AbstractHorse p_188514_2_)
     {
         EntitySkeleton entityskeleton = new EntitySkeleton(p_188514_2_.field_70170_p);
@@ -98,7 +104,7 @@
                 false
             )
         );
-        entityskeleton.field_70170_p.func_72838_d(entityskeleton);
+        if (!entityskeleton.field_70170_p.addEntity(entityskeleton, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY)) return null; // CraftBukkit
         return entityskeleton;
     }
 }
